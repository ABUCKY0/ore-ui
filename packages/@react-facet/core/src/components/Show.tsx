import React from 'react'
import { FacetProp, useFacetMap, useFacetWrap } from '@react-facet/core'
import { MountProps, NO_MOUNT_VALUE } from './Mount'
import { FastDivProps } from '@react-facet/dom-fiber'

type ShowProps<T> = MountProps<T> &
  Omit<FastDivProps, 'style'> & {
    display?: FacetProp<'flex' | 'none' | 'unset' | null | undefined>
  }

export const Show = <T,>({ when, children, display, is = true, not = NO_MOUNT_VALUE, ...props }: ShowProps<T>) => {
  const displayStyle = useFacetMap(
    (when, displayProp) => {
      const shouldShowChildren = not !== NO_MOUNT_VALUE ? when !== not : when === is
      return shouldShowChildren ? displayProp ?? 'unset' : 'none'
    },
    [is, not],
    [when, useFacetWrap(display)],
  )

  return (
    <fast-div style={{ display: displayStyle }} {...props}>
      <>{children}</>
    </fast-div>
  )
}
