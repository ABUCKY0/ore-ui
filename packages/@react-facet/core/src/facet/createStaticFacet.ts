import { Facet } from '../types'

export const createStaticFacet = <T>(value: T): Facet<T> => {
  return {
    get: () => value,
    observe: (listener) => {
      listener(value)
      return () => {}
    },
  }
}
