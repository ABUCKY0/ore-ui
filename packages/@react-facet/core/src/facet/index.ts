export * from './createFacet'
export * from './createReadOnlyFacet'
export * from './createStaticFacet'
export * from './wrapFacets'
