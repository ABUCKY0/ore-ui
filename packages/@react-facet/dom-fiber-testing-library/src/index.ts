import { setupAct } from './setupAct'
export * from './renderer'

setupAct()
