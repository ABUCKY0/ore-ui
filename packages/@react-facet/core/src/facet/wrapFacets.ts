import { FacetProp, Facet, isFacet } from '../types'
import { createStaticFacet } from './createStaticFacet'

export const wrapFacets = (facetProps: FacetProp<unknown>[]): Facet<unknown>[] => {
  return facetProps.map((facetProp) => (isFacet(facetProp) ? facetProp : createStaticFacet(facetProp)))
}
