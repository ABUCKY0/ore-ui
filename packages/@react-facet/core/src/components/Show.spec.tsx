import React from 'react'
import { render, act } from '@react-facet/dom-fiber-testing-library'
import { createFacet } from '../facet'
import { Show } from './Show'

it('shows children when true', () => {
  const displayFacet = createFacet({ initialValue: true })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Lorem ipsum</div>
  }

  const Example = () => {
    return (
      <Show when={displayFacet}>
        <Content />
      </Show>
    )
  }

  const scenario = <Example />

  const { container } = render(scenario)

  expect(rendered).toHaveBeenCalled()
  expect(container.firstChild).toHaveStyle('display: unset')
})

it('renders hidden children when false', () => {
  const displayFacet = createFacet({ initialValue: false })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Lorem ipsum</div>
  }

  const Example = () => {
    return (
      <Show when={displayFacet}>
        <Content />
      </Show>
    )
  }

  const scenario = <Example />

  const { container } = render(scenario)

  expect(rendered).toHaveBeenCalled()
  expect(container.firstChild).toHaveStyle('display: none')
})

it('can perform conditional rendering', () => {
  const numberFacet = createFacet<number>({ initialValue: 1 })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Lorem ipsum</div>
  }

  const Example = () => {
    return (
      <Show when={numberFacet} not={0}>
        <Content />
      </Show>
    )
  }

  const scenario = <Example />
  const { container } = render(scenario)

  expect(rendered).toHaveBeenCalled()
  expect(container.firstChild).toHaveStyle('display: unset')

  act(() => {
    numberFacet.set(0)
  })

  expect(container.firstChild).toHaveStyle('display: none')
})

it('can use FastDivProps other than style', () => {
  const stringFacet = createFacet<string | null>({ initialValue: 'text' })
  const classNameFacet = createFacet({ initialValue: 'some-class-name' })

  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Lorem ipsum</div>
  }

  const Example = () => {
    return (
      <Show when={stringFacet} not={''} className={classNameFacet}>
        <Content />
      </Show>
    )
  }

  const scenario = <Example />
  const { container } = render(scenario)

  expect(rendered).toHaveBeenCalled()
  expect(container.firstChild).toMatchSnapshot()
  expect(container.firstChild).toHaveStyle('display: unset')
  expect(container.firstChild).toHaveClass('some-class-name')
})

it('accepts a display prop', () => {
  const stringFacet = createFacet<string>({ initialValue: 'text' })

  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Lorem ipsum</div>
  }

  const Example = () => {
    return (
      <Show when={stringFacet} is={'text'} display={'flex'}>
        <Content />
      </Show>
    )
  }

  const scenario = <Example />
  const { container } = render(scenario)

  expect(rendered).toHaveBeenCalled()
  expect(container.firstChild).toHaveStyle('display: flex')
})
