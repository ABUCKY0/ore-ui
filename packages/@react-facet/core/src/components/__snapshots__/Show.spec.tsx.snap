// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can use FastDivProps other than style 1`] = `
<div
  class="some-class-name"
  style="display: unset;"
>
  <div>
    Lorem ipsum
  </div>
</div>
`;
